### PR TITLE
Fixed OpenVINO gemm layer

### LINKED ISSUE #25472: Test_ONNX_conformance.Layer_Test/test_gemm_default_matrix_bias_NGRAPH_CPU crashes
### System Information

OS: Ubuntu 18.04
OpenVINO: 2024.0

### Detailed description

Related PR: https://github.com/opencv/opencv/pull/25465

### Steps to reproduce

```
./bin/opencv_test_dnn --gtest_filter=Test_ONNX_conformance.Layer_Test/test_gemm_default_matrix_bias_NGRAPH_CPU
CTEST_FULL_OUTPUT
OpenCV version: 5.0.0-pre
OpenCV VCS version: 4.9.0-686-g07404e6da4-dirty
Build type: Release
Compiler: /usr/bin/c++  (ver 7.5.0)
Parallel framework: pthreads (nthreads=4)
CPU features: SSE SSE2 SSE3 *SSE4.1 *SSE4.2 *FP16? *AVX *AVX2? *AVX512-SKX?
Intel(R) IPP version: ippIP SSE4.2 (y8) 2021.11.0 (-) Feb  1 2024
Intel(R) IPP features code: 0x80
OpenCL Platforms: 
    NVIDIA CUDA
        dGPU: NVIDIA GeForce GTX 1080 (OpenCL 3.0 CUDA)
Current OpenCL device: 
    Type = dGPU
    Name = NVIDIA GeForce GTX 1080
    Version = OpenCL 3.0 CUDA
    Driver version = 530.30.02
    Address bits = 64
    Compute units = 20
    Max work group size = 1024
    Local memory size = 48 KB
    Max memory allocation size = 1 GB 1004 MB 16 KB
    Double support = Yes
    Half support = No
    Host unified memory = No
    Device extensions:
        cl_khr_global_int32_base_atomics
        cl_khr_global_int32_extended_atomics
        cl_khr_local_int32_base_atomics
        cl_khr_local_int32_extended_atomics
        cl_khr_fp64
        cl_khr_3d_image_writes
        cl_khr_byte_addressable_store
        cl_khr_icd
        cl_khr_gl_sharing
        cl_nv_compiler_options
        cl_nv_device_attribute_query
        cl_nv_pragma_unroll
        cl_nv_copy_opts
        cl_nv_create_buffer
        cl_khr_int64_base_atomics
        cl_khr_int64_extended_atomics
        cl_khr_device_uuid
        cl_khr_pci_bus_info
        cl_khr_external_semaphore
        cl_khr_external_memory
        cl_khr_external_semaphore_opaque_fd
        cl_khr_external_memory_opaque_fd
    Has AMD Blas = No
    Has AMD Fft = No
    Preferred vector width char = 1
    Preferred vector width short = 1
    Preferred vector width int = 1
    Preferred vector width long = 1
    Preferred vector width float = 1
    Preferred vector width double = 1
    Preferred vector width half = 0
TEST: Skip tests with tags: 'mem_6gb', 'verylong', 'dnn_skip_opencv_backend', 'dnn_skip_cpu', 'dnn_skip_cpu_fp16', 'dnn_skip_ocl', 'dnn_skip_ocl_fp16', 'dnn_skip_ie', 'dnn_skip_ie_ngraph', 'dnn_skip_ie_cpu', 'dnn_skip_ie_ocl', 'dnn_skip_ie_ocl_fp16', 'dnn_skip_onnx_conformance', 'dnn_skip_parser'
Note: Google Test filter = Test_ONNX_conformance.Layer_Test/test_gemm_default_matrix_bias_NGRAPH_CPU
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Test_ONNX_conformance
[ RUN      ] Test_ONNX_conformance.Layer_Test/test_gemm_default_matrix_bias_NGRAPH_CPU, where GetParam() = (test_gemm_default_matrix_bias, NGRAPH/CPU)
Ошибка сегментирования (стек памяти сброшен на диск)
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by alexlyulkov**
Fixed OpenVINO gemm layer
The problem was that our layer didn't properly handle all the possible gemm options in OpenVINO mode
Fixes #25472 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
The related ONNX conformance test was skipped in Abduragim's PR. Please enable it back.

**Comment by dkurt** [comment]
Please change a target branch to 4.x

**Comment by asmorkalov** [comment]
The PR brings several test regressions with OpenVINO 2022.1.1:
```
[  FAILED  ] 26 tests, listed below:
[  FAILED  ] Test_Model.Keypoints_face/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_Model.TextRecognitionWithCTCPrefixBeamSearch/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_layers.Linear/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_layers.Multiplication/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_layers.Constant/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_layers.LinearWithConstant/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_layers.MatmulWithTwoInputs/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Alexnet/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Googlenet/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.CaffeNet/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.RCNN_ILSVRC13/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.ZFNet/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.ResNet18v1/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.ResNet50v1/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.MobileNet_v2_FP16/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.LResNet100E_IR/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Inception_v2/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Inception_v1/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Shufflenet/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Resnet34_kinetics/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.ViT_B_32/0, where GetParam() = NGRAPH/CPU
```

**Comment by asmorkalov** [comment]
Regressions with OpenVINO 2024.0:
```
[  FAILED  ] Test_Int8_nets.Inception_v2/1, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_Int8_nets.Shufflenet/1, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_Model.Keypoints_face/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_Model.TextRecognition/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_Model.TextRecognitionWithCTCPrefixBeamSearch/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_layers.Linear/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_layers.Multiplication/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_layers.Constant/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_layers.LinearWithConstant/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_layers.MatmulWithTwoInputs/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Alexnet/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Googlenet/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.CaffeNet/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.RCNN_ILSVRC13/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.ZFNet/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.ResNet18v1/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.ResNet50v1/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.MobileNet_v2_FP16/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.LResNet100E_IR/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Inception_v2/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Inception_v1/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Shufflenet/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.Resnet34_kinetics/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] Test_ONNX_nets.ViT_B_32/0, where GetParam() = NGRAPH/CPU
```

**Comment by asmorkalov** [comment]
Failure example:
```
[----------] 1 test from Test_ONNX_nets
[ RUN      ] Test_ONNX_nets.Alexnet/0, where GetParam() = NGRAPH/CPU
unknown file: Failure
C++ exception with description "Check 'TRShape::broadcast_merge_into(output_shape, input_shapes[1], autob)' failed at src/core/shape_inference/include/eltwise_shape_inference.hpp:28:
While validating node 'opset1::Add Add_51 (opset1::MatMul MatMul_49[0]:f32[4096,1000], opset1::Constant Constant_50[0]:f32[1000,1]) -> (dynamic[...])' with friendly_name 'Add_51':
Argument shapes are inconsistent.
" thrown in the test body.
[  FAILED  ] Test_ONNX_nets.Alexnet/0, where GetParam() = NGRAPH/CPU (705 ms)
[----------] 1 test from Test_ONNX_nets (705 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (705 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] Test_ONNX_nets.Alexnet/0, where GetParam() = NGRAPH/CPU
```

File: modules/dnn/test/test_onnx_conformance_layer_parser_denylist.inl.hpp [hunk 1]
```diff
@@ -121,8 +121,8 @@
 "test_gemm_all_attributes",
 "test_gemm_alpha",
 "test_gemm_beta",
-"test_gemm_default_matrix_bias",
-"test_gemm_default_no_bias",
```
Comment (asmorkalov) [inline modules/dnn/test/test_onnx_conformance_layer_parser_denylist.inl.hpp:125]: Please remove.

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**